tracing-journald/tests/journal.rs

```diff
@@ -0,0 +1,145 @@
+#![cfg(target_os = "linux")]
+
+use std::collections::HashMap;
+use std::process::Command;
+
+use serde::Deserialize;
+use tracing::{error, info, warn, debug};
+use tracing_journald::Subscriber;
+use tracing_subscriber::subscribe::CollectExt;
+use tracing_subscriber::Registry;
+
+fn journalctl_version() -> std::io::Result<String> {
+    let output = Command::new("journalctl").arg("--version").output()?;
+    Ok(String::from_utf8_lossy(&output.stdout).to_string())
+}
+
+fn with_journald(f: impl FnOnce()) {
+    match journalctl_version() {
+        Ok(_) => {
+            let sub = Registry::default().with(Subscriber::new().unwrap().with_field_prefix(None));
+            tracing::collect::with_default(sub, f);
+        }
+        Err(error) => eprintln!(
+            "SKIPPING TEST: journalctl --version failed with error: {}",
+            error
+        ),
+    }
+}
+
+#[derive(Debug, PartialEq, Deserialize)]
+#[serde(untagged)]
+enum Field {
+    Text(String),
+    Binary(Vec<u8>)
+}
+
+// Convenience impls to compare fields against strings and bytes with assert_eq!
+impl PartialEq<&str> for Field {
+    fn eq(&self, other: &&str) -> bool {
+        match self {
+            Field::Text(s) => s == other,
+            Field::Binary(_) => false
+        }
+    }
+}
+
+impl PartialEq<[u8]> for Field {
+    fn eq(&self, other: &[u8]) -> bool {
+        match self {
+            Field::Text(s) => s.as_bytes() == other,
+            Field::Binary(data) => data == other
+        }
+    }
+}
+
+/// Read from journal with `journalctl`.
+///
+/// `test_name` is the randomized name of the test being run, and gets
+/// added as `TEST_NAME` match to the `journalctl` call, to make sure to
+/// only select journal entries originating from and relevant to the
+/// current test.
+fn read_from_journal(test_name: &str) -> Vec<HashMap<String, Field>> {
+    let stdout = String::from_utf8(
+        Command::new("journalctl")
+            .args(&["--user", "--output=json"])
+            // Filter by the PID of the current test process
+            .arg(format!("_PID={}", std::process::id()))
+            // tracing-journald logs strings in their debug representation
+            .arg(format!("TEST_NAME={:?}", test_name))
+            .output()
+            .unwrap()
+            .stdout,
+    )
+    .unwrap();
+
+    stdout
+        .lines()
+        .map(|l| {
+            dbg!(l);
+            serde_json::from_str(l).unwrap()
+        })
+        .collect()
```

Comment:
here's a potential solution to the flaky test failures:

my assumption is that the tests are likely failing occasionally because there may be a small delay in when we send the logs to `journald` and when they are actually written in the system journal. my guess is that this occurs more frequently on CI than on a dev machine because the CI containers typically have fewer cores than most people's dev machines, so it's possible that journald is less likely to be scheduled as frequently, or something. it seems like we can probably fix these flaky failures by adding a small number of retries to the `journalctl` command. 

i modified this code by adding a loop that will retry the `journalctl` command up to four times if the output is empty. after running the command five times, we can _probably_ assume that the logs actually weren't written to journald, so we return the empty vec. between retries, the test thread yields, in the hopes that this makes it more likely for the journald process to be scheduled.

```suggestion
    let mut cmd = Command::new("journalctl");
    cmd.args(&["--user", "--output=json"])
        // Filter by the PID of the current test process
        .arg(format!("_PID={}", std::process::id()))
        // tracing-journald logs strings in their debug representation
        .arg(format!("TEST_NAME={:?}", test_name));

    // Retry the `journalctl` command up to 5 times, in case there's some
    // timing-related flakiness.
    for retry in (0..5).rev() {
        eprintln!("running {:?} ({} tries remaining)...", cmd, retry);
        let stdout = String::from_utf8(cmd.output().unwrap().stdout).unwrap();

        let fields: Vec<HashMap<_, _>> = stdout
            .lines()
            .map(|l| {
                dbg!(l);
                serde_json::from_str(l).unwrap()
            })
            .collect();

        if !fields.is_empty() {
            return fields;
        }
    }

    Vec::new()
```

i've rerun the tests 3952 times so far without any failures...but i can't _guarantee_ this actually fixes it, of course.

Reply:
@hawkw Oh, many many thanks for testing this.  I think a retry is a good idea. I'll update the code accordingly, when I get back to working on this, perhaps with some little thread::sleep in between to not retry too fast.

Reply:
...oh, huh, I just saw it fail. sad!

<details>

<summary>Failure output</summary>

```

     Running tests/journal.rs (target/debug/deps/journal-4a6865a009a0bfc7)

running 4 tests
test internal_null_byte ... FAILED
test simple_message ... FAILED
test multiline_message_trailing_newline ... FAILED
test multiline_message ... FAILED

failures:

---- internal_null_byte stdout ----
running "journalctl" "--user" "--output=json" "_PID=2841561" "TEST_NAME=\"internal_null_byte\"" (4 tries remaining)...
running "journalctl" "--user" "--output=json" "_PID=2841561" "TEST_NAME=\"internal_null_byte\"" (3 tries remaining)...
running "journalctl" "--user" "--output=json" "_PID=2841561" "TEST_NAME=\"internal_null_byte\"" (2 tries remaining)...
running "journalctl" "--user" "--output=json" "_PID=2841561" "TEST_NAME=\"internal_null_byte\"" (1 tries remaining)...
running "journalctl" "--user" "--output=json" "_PID=2841561" "TEST_NAME=\"internal_null_byte\"" (0 tries remaining)...
thread 'internal_null_byte' panicked at 'assertion failed: `(left == right)`
  left: `0`,
 right: `1`', tracing-journald/tests/journal.rs:143:9
stack backtrace:
   0: rust_begin_unwind
             at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/std/src/panicking.rs:517:5
   1: core::panicking::panic_fmt
             at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/core/src/panicking.rs:101:14
   2: core::panicking::assert_failed_inner
   3: core::panicking::assert_failed
             at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/core/src/panicking.rs:140:5
   4: journal::internal_null_byte::{{closure}}
             at ./tests/journal.rs:143:9
   5: tracing_core::dispatch::with_default
             at /home/eliza/Code/tracing/tracing-core/src/dispatch.rs:263:5
   6: tracing::collect::with_default
             at /home/eliza/Code/tracing/tracing/src/collect.rs:24:5
   7: journal::with_journald
             at ./tests/journal.rs:21:13
   8: journal::internal_null_byte
             at ./tests/journal.rs:139:5
   9: journal::internal_null_byte::{{closure}}
             at ./tests/journal.rs:138:1
  10: core::ops::function::FnOnce::call_once
             at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/core/src/ops/function.rs:227:5
  11: core::ops::function::FnOnce::call_once
             at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/core/src/ops/function.rs:227:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

---- simple_message stdout ----
running "journalctl" "--user" "--output=json" "_PID=2841561" "TEST_NAME=\"simple_message\"" (4 tries remaining)...
running "journalctl" "--user" "--output=json" "_PID=2841561" "TEST_NAME=\"simple_message\"" (3 tries remaining)...
running "journalctl" "--user" "--output=json" "_PID=2841561" "TEST_NAME=\"simple_message\"" (2 tries remaining)...
running "journalctl" "--user" "--output=json" "_PID=2841561" "TEST_NAME=\"simple_message\"" (1 tries remaining)...
running "journalctl" "--user" "--output=json" "_PID=2841561" "TEST_NAME=\"simple_message\"" (0 tries remaining)...
thread 'simple_message' panicked at 'assertion failed: `(left == right)`
  left: `0`,
 right: `1`', tracing-journald/tests/journal.rs:98:9
stack backtrace:
   0: rust_begin_unwind
             at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/std/src/panicking.rs:517:5
   1: core::panicking::panic_fmt
             at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/core/src/panicking.rs:101:14
   2: core::panicking::assert_failed_inner
   3: core::panicking::assert_failed
             at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/core/src/panicking.rs:140:5
   4: journal::simple_message::{{closure}}
             at ./tests/journal.rs:98:9
   5: tracing_core::dispatch::with_default
             at /home/eliza/Code/tracing/tracing-core/src/dispatch.rs:263:5
   6: tracing::collect::with_default
             at /home/eliza/Code/tracing/tracing/src/collect.rs:24:5
   7: journal::with_journald
             at ./tests/journal.rs:21:13
   8: journal::simple_message
             at ./tests/journal.rs:94:5
   9: journal::simple_message::{{closure}}
             at ./tests/journal.rs:93:1
  10: core::ops::function::FnOnce::call_once
             at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/core/src/ops/function.rs:227:5
  11: core::ops::function::FnOnce::call_once
             at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/core/src/ops/function.rs:227:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

---- multiline_message_trailing_newline stdout ----
running "journalctl" "--user" "--output=json" "_PID=2841561" "TEST_NAME=\"multiline_message_trailing_newline\"" (4 tries remaining)...
running "journalctl" "--user" "--output=json" "_PID=2841561" "TEST_NAME=\"multiline_message_trailing_newline\"" (3 tries remaining)...
running "journalctl" "--user" "--output=json" "_PID=2841561" "TEST_NAME=\"multiline_message_trailing_newline\"" (2 tries remaining)...
running "journalctl" "--user" "--output=json" "_PID=2841561" "TEST_NAME=\"multiline_message_trailing_newline\"" (1 tries remaining)...
running "journalctl" "--user" "--output=json" "_PID=2841561" "TEST_NAME=\"multiline_message_trailing_newline\"" (0 tries remaining)...
thread 'multiline_message_trailing_newline' panicked at 'assertion failed: `(left == right)`
  left: `0`,
 right: `1`', tracing-journald/tests/journal.rs:129:9
stack backtrace:
   0: rust_begin_unwind
             at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/std/src/panicking.rs:517:5
   1: core::panicking::panic_fmt
             at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/core/src/panicking.rs:101:14
   2: core::panicking::assert_failed_inner
   3: core::panicking::assert_failed
             at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/core/src/panicking.rs:140:5
   4: journal::multiline_message_trailing_newline::{{closure}}
             at ./tests/journal.rs:129:9
   5: tracing_core::dispatch::with_default
             at /home/eliza/Code/tracing/tracing-core/src/dispatch.rs:263:5
   6: tracing::collect::with_default
             at /home/eliza/Code/tracing/tracing/src/collect.rs:24:5
   7: journal::with_journald
             at ./tests/journal.rs:21:13
   8: journal::multiline_message_trailing_newline
             at ./tests/journal.rs:122:5
   9: journal::multiline_message_trailing_newline::{{closure}}
             at ./tests/journal.rs:121:1
  10: core::ops::function::FnOnce::call_once
             at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/core/src/ops/function.rs:227:5
  11: core::ops::function::FnOnce::call_once
             at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/core/src/ops/function.rs:227:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

---- multiline_message stdout ----
running "journalctl" "--user" "--output=json" "_PID=2841561" "TEST_NAME=\"multiline_message\"" (4 tries remaining)...
running "journalctl" "--user" "--output=json" "_PID=2841561" "TEST_NAME=\"multiline_message\"" (3 tries remaining)...
running "journalctl" "--user" "--output=json" "_PID=2841561" "TEST_NAME=\"multiline_message\"" (2 tries remaining)...
running "journalctl" "--user" "--output=json" "_PID=2841561" "TEST_NAME=\"multiline_message\"" (1 tries remaining)...
running "journalctl" "--user" "--output=json" "_PID=2841561" "TEST_NAME=\"multiline_message\"" (0 tries remaining)...
thread 'multiline_message' panicked at 'assertion failed: `(left == right)`
  left: `0`,
 right: `1`', tracing-journald/tests/journal.rs:112:9
stack backtrace:
   0: rust_begin_unwind
             at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/std/src/panicking.rs:517:5
   1: core::panicking::panic_fmt
             at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/core/src/panicking.rs:101:14
   2: core::panicking::assert_failed_inner
   3: core::panicking::assert_failed
             at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/core/src/panicking.rs:140:5
   4: journal::multiline_message::{{closure}}
             at ./tests/journal.rs:112:9
   5: tracing_core::dispatch::with_default
             at /home/eliza/Code/tracing/tracing-core/src/dispatch.rs:263:5
   6: tracing::collect::with_default
             at /home/eliza/Code/tracing/tracing/src/collect.rs:24:5
   7: journal::with_journald
             at ./tests/journal.rs:21:13
   8: journal::multiline_message
             at ./tests/journal.rs:108:5
   9: journal::multiline_message::{{closure}}
             at ./tests/journal.rs:107:1
  10: core::ops::function::FnOnce::call_once
             at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/core/src/ops/function.rs:227:5
  11: core::ops::function::FnOnce::call_once
             at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/core/src/ops/function.rs:227:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.


failures:
    internal_null_byte
    multiline_message
    multiline_message_trailing_newline
    simple_message

test result: FAILED. 0 passed; 4 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.05s

error: test failed, to rerun pass '-p tracing-journald --test journal'
failed after 2487 runs!
```

</details>

but, this was with a version of the code without `thread::yield_now()` added, so perhaps that will help...you're probably correct in that this is related to flushing the send buffer?

Reply:
Maybe add a 100ms sleep after each failure?

Reply:
yeah, we could do that instead of _just_ yielding...

Reply:
So, yeah, turns out this wasn't as reliable as I though :pensive:  With your bash script I got the first failure after only a few dozen attempts :disappointed:   I'm very sorry for the extra roundtrips this caused, I should've caught this earlier.

I've added some retrying to the test: They now expect to read exactly one line and retry up to ten times if it fails.  There's a 100ms break between each retry, to avoid retrying "too fast".

Thanks a lot for your suggestion @hawkw.  I copied the idea but made a slightly different implementation with a separate `retry()` function.  Please do take another look :slightly_smiling_face: